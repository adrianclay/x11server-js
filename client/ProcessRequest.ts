export function _processPacket(processRequest: (dataView: DataView, ws: WebSocket) => void) {
  return function processPacket(dataView: DataView, ws: WebSocket) {
    const requestLengthInBytes = dataView.getInt16(2, true) * 4;
    processRequest(dataView, ws);
    if (requestLengthInBytes < dataView.byteLength) {
      processPacket(new DataView(dataView.buffer.slice(requestLengthInBytes)), ws);
    }
  }
}

type RequestHandler = (dataView : DataView, sequenceNumber: number, ws: WebSocket) => void;

export function _processRequest(
  handlers: { [opcode: number]: RequestHandler }
) {
  let sequenceNumber = 0;

  return function processRequest(dataView: DataView, ws: WebSocket) {
    const opCode = dataView.getInt8(0);
    sequenceNumber++;
    const handler = handlers[opCode];
    if(handler) {
      try {
        handler(dataView, sequenceNumber, ws);
      } catch (e) {
        // FixME: handle specific error types.
        // FixME: Only handle protocol defined errors.
        const nameError = (e.message === "Name" ? 15 : 7);
        const byteArray = new ArrayBuffer(32);
        const errorDataView = new DataView(byteArray);
        errorDataView.setUint8(0, 0);
        errorDataView.setUint8(1, nameError);
        errorDataView.setUint16(2, sequenceNumber, true);
        errorDataView.setUint8(10, opCode);

        ws.send(byteArray);
        console.log("Replaying an error to the server", e);
      }
    } else {
      console.error("Unhandled opCode: " + opCode)
    }
  }
}
