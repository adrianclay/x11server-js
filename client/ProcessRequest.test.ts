import {_processPacket, _processRequest} from "./ProcessRequest";

xtest('with both a CreateGC and GetProperty in 1 DataView', () => {
  const send = jest.fn();

  const bytes = new Uint8Array("3700060000000000020000000c000000ffffff00000000001400060002000000170000001f0000000000000000e1f505".match(/../g).map(s => Number.parseInt(s, 16)))
  _processRequest(new DataView(bytes.buffer), {send});

  expect(send).toBeCalledTimes(1);
});

describe('processPacket', () => {
  test('called with a packet containing 3 requests calls processRequest 3 times', () => {
    const processRequest = jest.fn();
    const processPacket = _processPacket(processRequest);
    const webSocket = jest.fn();

    const fakeRequest = [1, 1, 1, 0];

    const packetMadeOfThreeRequests = new Uint8Array([].concat(fakeRequest).concat(fakeRequest).concat(fakeRequest));
    processPacket(new DataView(packetMadeOfThreeRequests.buffer), webSocket);

    expect(processRequest).toBeCalledTimes(3);
  });
});

describe('processRequest', () => {
  test('It sends an error packet on error', () => {
    const webSocket = jest.fn();
    const processRequest = _processRequest({ 1: () => { throw new Error('Name'); } });
    const fakeRequest = [1, 1, 1, 0];
    processRequest(new DataView(new Uint8Array(fakeRequest).buffer), { send: webSocket });

    expect(webSocket).toBeCalledTimes(1);
    expect(new Uint8Array(webSocket.mock.calls[0][0])).toMatchSnapshot();
  })
});
