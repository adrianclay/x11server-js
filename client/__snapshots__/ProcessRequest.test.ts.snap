// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processRequest It sends an error packet on error 1`] = `
Uint8Array [
  0,
  15,
  1,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
]
`;
